File: android-native/polaris-study-kit/src/main/java/com/polaris/studykit/Readability.kt
package com.polaris.studykit

import kotlin.math.roundToInt

/**
 * Deterministic writing analysis for the "Essay" feature — no AI.
 * Flesch Reading Ease + Flesch–Kincaid grade level (1975 formulas) plus structural signals.
 */
data class ReadabilityReport(
    val words: Int,
    val sentences: Int,
    val syllables: Int,
    val fleschReadingEase: Double,
    val fleschKincaidGrade: Double,
    val avgSentenceLength: Double,
    val longSentences: Int,
    val passiveHits: Int,
) {
    fun summary(): String =
        "Reading ease ${fleschReadingEase.roundToInt()} · grade $fleschKincaidGrade · " +
            "$words words, $sentences sentences (avg $avgSentenceLength/sentence)"
}

object Readability {

    fun analyze(text: String): ReadabilityReport {
        val sentenceList = text.split(Regex("[.!?]")).map { it.trim() }.filter { it.isNotEmpty() }
        val wordList = text.split(Regex("[^A-Za-z']+")).filter { it.isNotEmpty() }

        val nWords = wordList.size.coerceAtLeast(1)
        val nSentences = sentenceList.size.coerceAtLeast(1)
        val nSyll = wordList.sumOf { syllables(it) }

        val wps = nWords.toDouble() / nSentences
        val spw = nSyll.toDouble() / nWords
        val ease = 206.835 - 1.015 * wps - 84.6 * spw
        val grade = 0.39 * wps + 11.8 * spw - 15.59

        val longS = sentenceList.count { it.split(" ").size > 25 }
        val passive = passiveHeuristic(wordList)

        return ReadabilityReport(
            words = nWords, sentences = nSentences, syllables = nSyll,
            fleschReadingEase = round1(ease), fleschKincaidGrade = round1(grade),
            avgSentenceLength = round1(wps), longSentences = longS, passiveHits = passive,
        )
    }

    /** Heuristic English syllable count (vowel-group based, silent-'e' adjustment). */
    fun syllables(word: String): Int {
        val w = word.lowercase().filter { it.isLetter() }
        if (w.isEmpty()) return 0
        val vowels = "aeiouy".toSet()
        var count = 0
        var prevVowel = false
        for (ch in w) {
            val isVowel = ch in vowels
            if (isVowel && !prevVowel) count++
            prevVowel = isVowel
        }
        if (w.endsWith("e") && count > 1) count--
        return count.coerceAtLeast(1)
    }

    private fun passiveHeuristic(words: List<String>): Int {
        val be = setOf("is", "are", "was", "were", "be", "been", "being")
        var hits = 0
        for (i in 0 until (words.size - 1).coerceAtLeast(0)) {
            if (words[i].lowercase() in be && words[i + 1].lowercase().endsWith("ed")) hits++
        }
        return hits
    }

    private fun round1(x: Double): Double = (x * 10).roundToInt() / 10.0
}

File: ios-native/Sources/PolarisStudyKit/Readability.swift
import Foundation

/// Deterministic writing analysis for the "Essay" feature — no AI.
///
/// Computes **Flesch Reading Ease** and the **Flesch–Kincaid grade level** (classic 1975
/// formulas) plus simple structural signals (avg sentence length, long-sentence and
/// passive-voice heuristics) to give students concrete, rule-based feedback.
public struct ReadabilityReport: Sendable {
    public let words: Int
    public let sentences: Int
    public let syllables: Int
    public let fleschReadingEase: Double   // higher = easier (0–100+)
    public let fleschKincaidGrade: Double  // US grade level
    public let avgSentenceLength: Double
    public let longSentences: Int          // > 25 words
    public let passiveHits: Int            // heuristic

    public var summary: String {
        String(format: "Reading ease %.0f · grade %.1f · %d words, %d sentences (avg %.1f/sentence)",
               fleschReadingEase, fleschKincaidGrade, words, sentences, avgSentenceLength)
    }
}

public enum Readability {

    public static func analyze(_ text: String) -> ReadabilityReport {
        let sentenceList = text.split(whereSeparator: { ".!?".contains($0) })
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
        let wordList = text.split(whereSeparator: { !$0.isLetter && $0 != "'" }).map(String.init)

        let nWords = max(1, wordList.count)
        let nSentences = max(1, sentenceList.count)
        let nSyll = wordList.reduce(0) { $0 + syllables(in: $1) }

        let wps = Double(nWords) / Double(nSentences)
        let spw = Double(nSyll) / Double(nWords)
        let ease = 206.835 - 1.015 * wps - 84.6 * spw
        let grade = 0.39 * wps + 11.8 * spw - 15.59

        let longS = sentenceList.filter { $0.split(separator: " ").count > 25 }.count
        let passive = passiveHeuristic(wordList)

        return ReadabilityReport(
            words: nWords, sentences: nSentences, syllables: nSyll,
            fleschReadingEase: round(ease * 10) / 10,
            fleschKincaidGrade: round(grade * 10) / 10,
            avgSentenceLength: round(wps * 10) / 10,
            longSentences: longS, passiveHits: passive)
    }

    /// Heuristic English syllable count (vowel-group based, with common adjustments).
    static func syllables(in word: String) -> Int {
        let w = word.lowercased().filter { $0.isLetter }
        guard !w.isEmpty else { return 0 }
        let vowels = Set("aeiouy")
        var count = 0
        var prevVowel = false
        for ch in w {
            let isVowel = vowels.contains(ch)
            if isVowel && !prevVowel { count += 1 }
            prevVowel = isVowel
        }
        if w.hasSuffix("e") && count > 1 { count -= 1 }         // silent 'e'
        return max(1, count)
    }

    /// Count "to be" + past-participle-ish patterns as a rough passive-voice signal.
    private static func passiveHeuristic(_ words: [String]) -> Int {
        let be: Set<String> = ["is", "are", "was", "were", "be", "been", "being"]
        var hits = 0
        for i in 0..<max(0, words.count - 1) {
            if be.contains(words[i].lowercased()), words[i + 1].lowercased().hasSuffix("ed") {
                hits += 1
            }
        }
        return hits
    }
}
